File: training-data/zig/quickselect.zig
const std = @import("std");

fn partition(items: []i32) usize {
    const pivot = items[items.len - 1];
    var i: usize = 0;
    var j: usize = 0;
    while (j < items.len - 1) : (j += 1) {
        if (items[j] < pivot) {
            const tmp = items[i];
            items[i] = items[j];
            items[j] = tmp;
            i += 1;
        }
    }
    const tmp = items[i];
    items[i] = items[items.len - 1];
    items[items.len - 1] = tmp;
    return i;
}

/// Returns the k-th smallest element (0-indexed) via Lomuto-partition quickselect.
fn quickselect(items: []i32, k: usize) i32 {
    if (items.len == 1) return items[0];
    const p = partition(items);
    if (k == p) return items[p];
    if (k < p) return quickselect(items[0..p], k);
    return quickselect(items[p + 1 ..], k - p - 1);
}

pub fn main() void {
    var data = [_]i32{ 7, 2, 1, 6, 8, 5, 3, 4 };
    const fourth_smallest = quickselect(&data, 3);
    std.debug.print("{d}\n", .{fourth_smallest});

    var data2 = [_]i32{ 9, 1, 5, 3 };
    const median_low = quickselect(&data2, 1);
    std.debug.print("{d}\n", .{median_low});
}

File: training-data/python/quickselect.py
import random


def quickselect(items, k):
    """Return the k-th smallest element (0-indexed) in expected O(n) time."""
    items = list(items)

    def select(lo, hi, k):
        if lo == hi:
            return items[lo]
        pivot_index = random.randint(lo, hi)
        pivot_index = partition(lo, hi, pivot_index, k)
        if k == pivot_index:
            return items[k]
        elif k < pivot_index:
            return select(lo, pivot_index - 1, k)
        else:
            return select(pivot_index + 1, hi, k)

    def partition(lo, hi, pivot_index, k):
        pivot_value = items[pivot_index]
        items[pivot_index], items[hi] = items[hi], items[pivot_index]
        store_index = lo
        for i in range(lo, hi):
            if items[i] < pivot_value:
                items[store_index], items[i] = items[i], items[store_index]
                store_index += 1
        items[hi], items[store_index] = items[store_index], items[hi]
        return store_index

    return select(0, len(items) - 1, k)


if __name__ == "__main__":
    data = [7, 10, 4, 3, 20, 15]
    for k in range(len(data)):
        print(f"{k}-th smallest:", quickselect(data, k))
